server/player/player.go

```diff
@@ -957,6 +980,23 @@ func (p *Player) respawn(f func(p *Player)) {
 	})
 }
 
+// spawnLocation returns position and world where player should be spawned.
+func (p *Player) spawnLocation() (playerSpawn cube.Pos, w *world.World, spawnBlockBroken bool, previousDimension world.Dimension) {
+	tx := p.tx
+	w = tx.World()
+	previousDimension = w.Dimension()
+	playerSpawn = w.PlayerSpawn(p.UUID())
+	if b, ok := tx.Block(playerSpawn).(block.RespawnBlock); ok && b.CanRespawnOn() {
+		return playerSpawn, w, false, previousDimension
+	}
+
+	// We can use the principle here that returning through a portal of a specific dimension inside that dimension will
+	// always bring us back to the overworld.
+	w = w.PortalDestination(w.Dimension())
+	worldSpawn := w.Spawn()
+	return worldSpawn, w, playerSpawn != worldSpawn, previousDimension
+}
+
```

Comment:
This function does not return accurate spawn positions. block.RespawnBlock should have a method which returns the position where the player should respawn since it's not at the same place where the bed/respawn anchor is located.

https://minecraft.wiki/w/Bed#Setting_the_spawn_point
https://minecraft.wiki/w/Respawn_Anchor#Respawning